File: java/core/src/main/java/sleeper/core/range/Region.java
/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.range;

import sleeper.core.key.Key;
import sleeper.core.schema.Schema;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * A Region is a rectangular area in the space of row keys. This is created by
 * specifying a {@link Range} for each row key dimension. (If a range is not
 * specified for a dimension then it is implicitly assumed to cover the entire
 * space in that dimension.)
 */
public class Region {
    private final Map<String, Range> rowKeyFieldNameToRange;

    public Region(Collection<Range> ranges) {
        rowKeyFieldNameToRange = new HashMap<>();
        for (Range range : ranges) {
            if (rowKeyFieldNameToRange.containsKey(range.getFieldName())) {
                throw new IllegalArgumentException("Should only provide one range per row key field");
            }
            rowKeyFieldNameToRange.put(range.getFieldName(), range);
        }
    }

    public Region(Range range) {
        this(Map.of(range.getFieldName(), range));
    }

    private Region(Map<String, Range> rowKeyFieldNameToRange) {
        this.rowKeyFieldNameToRange = rowKeyFieldNameToRange;
    }

    /**
     * Gets the range for a field.
     *
     * @param  fieldName the field name
     * @return           the range for the field
     */
    public Range getRange(String fieldName) {
        return rowKeyFieldNameToRange.get(fieldName);
    }

    /**
     * Returns the ranges in this region unordered.
     *
     * <strong>Note: these ranges are returned in no
     * particular order.</strong>
     *
     * @return a List of Ranges in no particular order.
     * @see    Region#getRangesOrdered(Schema)
     */
    public Collection<Range> getRangesUnordered() {
        return Collections.unmodifiableCollection(rowKeyFieldNameToRange.values());
    }

    /**
     * Returns the ranges in this region ordered according to the given schema.
     *
     * All row key dimensions in the schema must be present in this region,
     * otherwise an exception is thrown.
     *
     * @param  schema                   the schema that dictates returned order
     * @return                          a List of Ranges in schema order.
     * @throws IllegalArgumentException if a row key in the given schema is not present in this region
     * @see                             getPartialRangesOrdered
     */
    public List<Range> getRangesOrdered(Schema schema) {
        List<String> rowKeysOrdered = schema.getRowKeyFieldNames();
        List<Range> ranges = new ArrayList<>(rowKeysOrdered.size());
        for (String rowKey : rowKeysOrdered) {
            Range range = getRange(rowKey);
            if (range == null) {
                throw new IllegalArgumentException("schema row key \"" + rowKey + "\" does not exist in this Region");
            }
            ranges.add(range);
        }
        return ranges;
    }

    /**
     * Returns the ranges in this region ordered according to the given schema.
     * If a given row key dimension is not present in this region, it is skipped.
     *
     * @param  schema the schema that dictates returned order
     * @return        a List of Ranges in schema order.
     * @see           getRangesOrdered
     */
    public List<Range> getPartialRangesOrdered(Schema schema) {
        List<String> rowKeysOrdered = schema.getRowKeyFieldNames();
        List<Range> ranges = new ArrayList<>(rowKeysOrdered.size());
        for (String rowKey : rowKeysOrdered) {
            Range range = getRange(rowKey);
            if (range != null) {
                ranges.add(range);
            }
        }
        return ranges;
    }

    /**
     * Returns the column indexes of row key columns from the schema that have a corresponding range in this region.
     *
     * For each row key column in schema order, its column index is included in the returned list if this region
     * contains a range for that column.
     *
     * @param  schema the schema to use for ordering and field names
     * @return        a list of indexes for row key fields present in this region
     */
    public List<Integer> getRowKeyColumnIndexesInRegion(Schema schema) {
        List<Integer> indexes = new ArrayList<>();
        List<String> rowKeysOrdered = schema.getRowKeyFieldNames();
        for (int i = 0; i < rowKeysOrdered.size(); i++) {
            if (getRange(rowKeysOrdered.get(i)) != null) {
                indexes.add(i);
            }
        }
        return indexes;
    }

    /**
     * Checks whether the provided key is contained within this region.
     *
     * @param  schema the Sleeper schema
     * @param  key    the key to check
     * @return        whether the key is in this range
     */
    public boolean isKeyInRegion(Schema schema, Key key) {
        if (null == key || key.isEmpty()) {
            throw new IllegalArgumentException("Key must be non-null and not empty");
        }

        // Key is not in region if any dimension is not in the corresponding range.
        int i = 0;
        for (Object object : key.getKeys()) {
            String rowKeyFieldName = schema.getRowKeyFields().get(i).getName();
            if (rowKeyFieldNameToRange.containsKey(rowKeyFieldName)) {
                Range range = rowKeyFieldNameToRange.get(rowKeyFieldName);
                if (!range.doesRangeContainObject(object)) {
                    return false;
                }
            }
            i++;
        }

        return true;
    }

    /**
     * Checks whether the provided region overlaps with this region.
     *
     * @param  otherRegion the region to check
     * @return             whether the region overlaps with this region
     */
    public boolean doesRegionOverlap(Region otherRegion) {
        // Two regions overlap if they overlap in all dimensions
        for (Map.Entry<String, Range> entry : rowKeyFieldNameToRange.entrySet()) {
            Range otherRange = otherRegion.rowKeyFieldNameToRange.get(entry.getKey());
            if (null != otherRange) {
                if (!entry.getValue().doesRangeOverlap(otherRange)) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Creates a copy of this region which contains the provided range. Note that if there is an existing range for the
     * same field, it will be replaced by the supplied range.
     *
     * @param  range the range to add to the copy
     * @return       a copy of this region which contains the provided range
     */
    public Region copyWithRange(Range range) {
        Map<String, Range> newRanges = new HashMap<>(rowKeyFieldNameToRange);
        newRanges.put(range.getFieldName(), range);
        return new Region(newRanges);
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 59 * hash + Objects.hashCode(this.rowKeyFieldNameToRange);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        Region other = (Region) obj;
        if (!Objects.equals(this.rowKeyFieldNameToRange, other.rowKeyFieldNameToRange)) {
            return false;
        }
        return true;
    }

    @Override
    public String toString() {
        return "Region{rowKeyFieldNameToRange=" + rowKeyFieldNameToRange + '}';
    }
}

File: rust/sleeper_core/src/datafusion/region.rs
//! Functions for handling file regions. That is a range of key values, e.g. from 'b' - 'h'.
/*
* Copyright 2022-2025 Crown Copyright
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
use datafusion::logical_expr::{Expr, ident, lit};
use std::collections::HashMap;

/// Represents a Sleeper partition region.
///
/// A [`SleeperPartitionRegion`] is multi-dimension key range over row-key fields in Sleeper.
/// If a table has only on row-key field then a region is a single row range. A region in a
/// table with two row-key fields would be a rectangle, etc.
#[derive(Debug, Default)]
pub struct SleeperRegion<'a> {
    pub region: HashMap<String, ColRange<'a>>,
}

/// Defines a partition range of a single field.
#[derive(Debug, Copy, Clone)]
pub struct ColRange<'a> {
    pub lower: PartitionBound<'a>,
    pub lower_inclusive: bool,
    pub upper: PartitionBound<'a>,
    pub upper_inclusive: bool,
}

/// Type safe variant for Sleeper partition boundary
#[derive(Debug, Copy, Clone)]
pub enum PartitionBound<'a> {
    Int32(i32),
    Int64(i64),
    String(&'a str),
    ByteArray(&'a [u8]),
    /// Represented by a NULL in Java
    Unbounded,
}

impl<'a> SleeperRegion<'a> {
    /// Create new region.
    #[must_use]
    pub fn new(region: HashMap<String, ColRange<'a>>) -> Self {
        Self { region }
    }

    /// Number of dimensions in region.
    #[must_use]
    pub fn len(&self) -> usize {
        self.region.len()
    }

    /// Empty check
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.region.is_empty()
    }
}

/// Create the `DataFusion` filtering expression from a Sleeper region.
///
/// For each field in the row keys, we look up the partition range for that
/// field and create a expression tree that combines all the various filtering conditions.
impl From<&SleeperRegion<'_>> for Option<Expr> {
    fn from(value: &SleeperRegion<'_>) -> Self {
        let mut col_expr: Option<Expr> = None;
        for (name, range) in &value.region {
            let lower_expr = lower_bound_expr(range, name);
            let upper_expr = upper_bound_expr(range, name);
            let expr = match (lower_expr, upper_expr) {
                (Some(l), Some(u)) => Some(l.and(u)),
                (Some(l), None) => Some(l),
                (None, Some(u)) => Some(u),
                (None, None) => None,
            };
            // Combine this column filter with any previous column filter
            if let Some(e) = expr {
                col_expr = match col_expr {
                    Some(original) => Some(original.and(e)),
                    None => Some(e),
                }
            }
        }
        col_expr
    }
}

/// Calculate the upper bound expression on a given [`ColRange`].
///
/// This takes into account the inclusive/exclusive nature of the bound.
///
fn upper_bound_expr(range: &ColRange, name: &String) -> Option<Expr> {
    if let PartitionBound::Unbounded = range.upper {
        None
    } else {
        let max_bound = bound_to_lit_expr(&range.upper);
        if range.upper_inclusive {
            Some(ident(name).lt_eq(max_bound))
        } else {
            Some(ident(name).lt(max_bound))
        }
    }
}

/// Calculate the lower bound expression on a given [`ColRange`].
///
/// Not all bounds are present, so `None` is returned for the unbounded case.
///
/// This takes into account the inclusive/exclusive nature of the bound.
///
fn lower_bound_expr(range: &ColRange, name: &String) -> Option<Expr> {
    if let PartitionBound::Unbounded = range.lower {
        None
    } else {
        let min_bound = bound_to_lit_expr(&range.lower);
        if range.lower_inclusive {
            Some(ident(name).gt_eq(min_bound))
        } else {
            Some(ident(name).gt(min_bound))
        }
    }
}

/// Convert a [`PartitionBound`] to an [`Expr`] that can be
/// used in a bigger expression.
///
/// # Panics
/// If bound is [`PartitionBound::Unbounded`] as we can't construct
/// an expression for that.
///
fn bound_to_lit_expr(bound: &PartitionBound) -> Expr {
    match bound {
        PartitionBound::Int32(val) => lit(*val),
        PartitionBound::Int64(val) => lit(*val),
        PartitionBound::String(val) => lit(val.to_owned()),
        PartitionBound::ByteArray(val) => lit(val.to_owned()),
        PartitionBound::Unbounded => {
            panic!("Can't create filter expression for unbounded partition range!");
        }
    }
}
